### PR TITLE
Improved 'How to contribute' and contribution guidelines

### PR DESCRIPTION
**Pull request by fhueske**
I updated the "How to contribute" page and added guides for contributing code, documentation, and website improvements. Coding guidelines should be updated by need as discussed on the mailing list.

Please review and give feedback.

### CODE REVIEW

**Comment by mxm** [comment]
+1 you nicely reworked the main how-to-contribute page into three distinct pages.

File: contribute-code.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
+---
+title:  "Contributing Code"
+---
+
+Apache Flink is maintained, improved, and extended by code contributions of volunteers. The Apache Flink community encourages anybody to contribute source code. In order to ensure a pleasant contribution experience for contributors and reviewers and to preserve the high quality of the code base, we follow a contribution process that is explained in this document.
+
+This document contains everything you need to know about contributing code to Apache Flink. It describes the process of preparing, testing and submitting a contributing, explains coding guidelines and code style of Flink's code base, and gives instructions to setup a development environment.
```
Comment (vasia) [inline contribute-code.md]: testing and submitting a *contribution

File: contribute-code.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+If the description of a JIRA issue indicates that its resolution will touch sensible parts of the code base, be sufficiently complex, or add significant amounts of new code, the Flink community might requests a design document (most contributions should not require a design document). The purpose of this document is to ensure that the overall approach to address the issue is sensible and agreed upon by the community. JIRA issues that require a design document are tagged with the **`requires-design-doc`** label. The label can be attached by any community member who feels that a design document is necessary. A good description helps to decide whether a JIRA issue requires a design document or not. The design document must be added or attached to or link from the JIRA issue and cover the following aspects:
```
Comment (vasia) [inline contribute-code.md]: the Flink community might *request

File: contribute-code.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+- **Commit message**. A pull request must relate to a JIRA issue; create an issue if none exists for the change you want to make. The latest commit message should reference that issue. An example commit message would be *[FLINK-633] Fix NullPointerException for empty UDF parameters*. That way, the pull request automatically gives a description of what it does, for example what bug does it fix in what way.
+
+- **Append review commits**. When you get comments on the pull request asking for changes, append commits for these changes. *Do not rebase and squash them.* It allows people to review the cleanup work independently. Otherwise reviewers have to go through the entire set of diffs again.
+
+### Exceptions and error messages
+{:.no_toc} 
+
+- **Exception swallowing**. Do not swallow exceptions and print the stacktrace. Instead check how exceptions are handled in similar situations.
```
Comment (vasia) [inline contribute-code.md]: Could we maybe add an example of a "similar situation" here?

**Comment by vasia** [comment]
Thank you @fhueske! It looks good much better now :+1:

File: contribute-code.md (hunk 1 above)
Comment (mjsax) [inline contribute-code.md]: preparing, testing COMMA and

File: contribute-code.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+- **Meaningful error messages**. Give meaningful exception messages. Try to imagine why an exception could be thrown (what a user did wrong) and give a message that will help a user to resolve the problem.
+
+### Tests
+{:.no_toc} 
+
+- **Tests need to pass**. Any pull request where the tests do not pass or which does not compile will not undergo any further review. We recommend to connect your private GitHub accounts with [Travis CI](http://travis-ci.org/) (like the Flink GitHub repository). Travis will run tests for all tested environments whenever you push something into *your* Github repository. Please note the previous [comment about flaky tests]( {{site.base}}/contribute-code.html#verifying-the-compliance-of-your-code).
+
```
Comment (mjsax) [inline contribute-code.md]: We should add a "Test required for bug fixes" paragraph (or extend the paragraph "Tests for new features are required")

File: contribute-code.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+### Documentation
+{:.no_toc}
+
+- **Documentation Updates**. Many changes in the system will also affect the documentation (both JavaDocs and the user documentation in the `docs/` directory.). Pull requests and patches are required to update the documentation accordingly, otherwise the change can not be accepted to the source code. See the [Contribute documentation]({{site.base}}/contribute-documentation.html) guide for how to update the documentation.
+
+- **Javadocs for public methods**. Public methods and classes that are part of the user-facing API need to have JavaDocs. Please write meaningful docs. Good docs are concise and informative.
```
Comment (mjsax) [inline contribute-code.md]: Why only "user-facing API" ? IMHO, all public methods (except trival getter/setter) should be documented.

File: how-to-contribute.md [hunk 6]
```diff
@@ -1,191 +1,147 @@
...
-```
+## Help others and join the discussions
 
-3. It is typically helpful to switch to a *topic branch* for the changes. To create a dedicated branch based on the current master, use the following command:
+Most communication in the Apache Flink community happens on two mailing lists:
 
-    ```
-git checkout -b myBranch master
-```
+- The user mailing lists `user@flink.apache.org` is the place where users of Apache Flink ask questions and seek for help or advice. Joining the user list and helping other users is a very good way to contribute to Flink's community. Furthermore, there is the [[Flink]](http://stackoverflow.com/questions/tagged/flink) tag on Stack Overflow if you'd like to help Flink users (and harvest some points) there.
```
Comment (mjsax) [inline how-to-contribute.md]: The SO tags got merged with "apche-flink" being the main tag -> "https://stackoverflow.com/questions/tagged/apache-flink"

Comment (fhueske) [inline how-to-contribute.md]: The `flink` tag was there first and had more questions, `apache-flink` was added later. That's why we communicated `flink`. Cool they got merged :-)

File: contribute-code.md (hunk 4 above)
Comment (fhueske) [inline contribute-code.md]: I just copied the existing coding guidelines. I suggest you start a discussion about this on the ML and we can add it later. I would not block the PR on that.

File: how-to-contribute.md (hunk 6 above)
Comment (mjsax) [inline how-to-contribute.md]: There is no "flink" tag anymore... it always redirect to "apache-flink"; all old questions are automatically retaged and there is just a single count for all question now (which is the sum of "flink" and "apache-flink")

File: contribute-code.md (hunk 3 above)
Comment (fhueske) [inline contribute-code.md]: That would depend on the code where the exception should be handled, e.g., a test, the API, runtime, etc.
But I agree, I will try to rephrase the statement.

File: how-to-contribute.md (hunk 6 above)
Comment (fhueske) [inline how-to-contribute.md]: Yes, I saw that. But that's quite new, I think.

**Comment by fhueske** [comment]
Thanks for the review. I'll addressed your comments.

File: contribute-documentation.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+The Flink documentation is written in [Markdown](http://daringfireball.net/projects/markdown/). Markdown is a lightweight markup language which can be translated to HTML.
+
+In order to update or extend the documentation you have to modify the Markdown (`.md`) files. Please verify your changes by starting the build script in preview mode.
+
+```
+cd docs
+./build_docs.sh -p
+```
+
+The script compiles the Markdown files into static HTML pages and starts a local webserver. Open your browser at `http://localhost:4000` to view the compiled documentation including your changes. The served documentation is automatically re-compiled and updated when you modify and safe Markdown files and refresh your browser.
```
Comment (chiwanpark) [inline contribute-documentation.md]: Typo: when you modify and **save** Markdown files

**Comment by fhueske** [comment]
Thanks, will fix the typo and merge this PR.

**Pull request closed without merging**